[blog] Add reflection blog (#4263)
pulumi/docs: content/blog/reflections-of-a-pulumi-intern/index.md

---
title: "Reflections of a Pulumi Intern"

> meta_image: meta.png

Reviewer: @spara, should we copy and use the same meta image used in Sashu's post?

Reviewer: Looks like Sophia is out today. Let's use the same meta image Sashu used.

authors: ["albert-zhong"]

> meta_desc: "A reflection of my Pulumi intern experiences"

Reviewer: The linter is complaining: 

```
Meta description is too short. Must be at least 50 characters.
```

tags: ["pulumi", "intern"]
date: "2020-09-18"
---

Hi everyone! I'm Albert, a soon-to-be sophomore studying computer science at the University of Washington. Today marks my last day as a Pulumi intern, so I figure I'd reflect on my experiences up until this point.

## Joining Pulumi

I heard of Pulumi for the first time when they visited my school's career fair in January. As I rounded the corner into the [CSE1 atrium](https://www.cs.washington.edu/building), I saw a banner titled "Modern Infrastructure as Code," with some lines of TypeScript of what appeared to be the creation of an S3 bucket. I found it interesting because it looked so simple and modular. In contrast, when I had played around with AWS in the past to deploy the ocassional Django web app, it had always involved following a series of exact instructions (usually go to this dashboard, copy this YAML file, run this command, and pray it works).

> I walked up to the banner's adjacent booth and chatted with [Sean](https://www.pulumi.com/blog/author/sean-holung/), a software engineer at Pulumi. He explained how Pulumi was building an open-source platform that allows developers to manage cloud infrastructure using familiar programming languages. To be honest I didn't have too much experience with the cloud, but I did sympathize with the pain of wrangling with YAML files. The concept sounded compelling, so I told him I was interested in interning there. He gave me the Pulumi recruiter's email, who I later emailed that day with my resume and a brief bit about my background. After a few email and phone exchanges, I came over to the Pulumi office for an in-person interview (which was unforunately also the only time I ever saw it, due to [COVID-19](https://www.pulumi.com/blog/coronavirus-plan/)).

Reviewer: The diff doesn't clearly show the changes, but it's just updating the two links:

 - `[Sean](/blog/author/sean-holung/)`
 - `[COVID-19]({{< relref "/blog/coronavirus-plan" >}})`

`relref` ensures the page exists when Hugo builds the site. Would love to be able to use it for the author pages, but it doesn't work for those, so just use relative links on their own.

```suggestion
I walked up to the banner's adjacent booth and chatted with [Sean](/blog/author/sean-holung/), a software engineer at Pulumi. He explained how Pulumi was building an open-source platform that allows developers to manage cloud infrastructure using familiar programming languages. To be honest I didn't have too much experience with the cloud, but I did sympathize with the pain of wrangling with YAML files. The concept sounded compelling, so I told him I was interested in interning there. He gave me the Pulumi recruiter's email, who I later emailed that day with my resume and a brief bit about my background. After a few email and phone exchanges, I came over to the Pulumi office for an in-person interview (which was unforunately also the only time I ever saw it, due to [COVID-19]({{< relref "/blog/coronavirus-plan" >}})).
```

## The First Week

I had very little idea what the internship would look like. In addition to it being virtual, it was my first software engineering gig, so I had no prior experience to base it off. I wasn't sure what my role would be—would I just be assigned a single project and be told to work on it? Would I have a mentor? How does the process of submitting code look like? Fortunately, these questions were answered fairly quickly.

For the first week, the other interns and I converted Pulumi [examples](https://github.com/pulumi/examples/) from one language to another. TypeScript seemed to be the most common language, so I converted a few from that language to some of the relatively less popular languages, such as Python and Go. I thought that this was an excellent way to help us interns ramp up, since it allowed us to get some baseline experience with Pulumi and the various cloud providers and programming languages it supports. By Friday, I was pretty familiar with how to navigate the various Pulumi repositories and documentation, which was a huge benefit as I launched into my project.

## My Project: `crd2pulumi`

> On the following Monday, I joined a Zoom meeting with [Levi](https://www.pulumi.com/blog/author/levi-blackstone/) and my mentor [Justin](https://www.pulumi.com/blog/author/justin-vanpatten/). They had written up a preliminary design doc on generating SDK types for Kubernetes CustomResources, which would be my project for the summer. I had little experiences with Kubernetes, so I couldn't fully conceptualize what the end product would look like until a few days later.

Reviewer: Prefer relative links:

```suggestion
On the following Monday, I joined a Zoom meeting with [Levi](/blog/author/levi-blackstone/) and my mentor [Justin](/blog/author/justin-vanpatten/). They had written up a preliminary design doc on generating SDK types for Kubernetes CustomResources, which would be my project for the summer. I had little experiences with Kubernetes, so I couldn't fully conceptualize what the end product would look like until a few days later.
```

> Basically, you can define your own fields of a Kubernetes CustomResource using [CustomResourceDefinitions](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/) (CRDs). Pulumi already allows you to create CustomResources using the [apiextensions](https://www.pulumi.com/docs/reference/pkg/kubernetes/apiextensions/) package. However there's no way to inherently know the exact fields of a CustomResource, since it's set by some CRD. Therefore we have to set the CustomResource fields to some generic type, such as `map[string]interface{}` in Go or `pulumi.Input<any>` in TypeScript. This makes filling out the fields for for complex CRDs such as cert-manager or Istio very annoying, since they contain thousands of nested properties. Being able to leverage IDE type-checking and auto-complete makes the process much quicker and less error-prone.

Reviewer: Update link to use `relref`

```suggestion
Basically, you can define your own fields of a Kubernetes CustomResource using [CustomResourceDefinitions](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/) (CRDs). Pulumi already allows you to create CustomResources using the [apiextensions]({{< relref "/docs/reference/pkg/kubernetes/apiextensions" >}}) package. However there's no way to inherently know the exact fields of a CustomResource, since it's set by some CRD. Therefore we have to set the CustomResource fields to some generic type, such as `map[string]interface{}` in Go or `pulumi.Input<any>` in TypeScript. This makes filling out the fields for for complex CRDs such as cert-manager or Istio very annoying, since they contain thousands of nested properties. Being able to leverage IDE type-checking and auto-complete makes the process much quicker and less error-prone.
```

> Over the course of the summer, I progessively built `crd2pulumi`, a CLI tool that generates a typed CustomResource from a CRD. You can find the latest releases [here](https://github.com/pulumi/crd2pulumi). I then leveraged this tool to generate libraries for various Kubernetes Operators, which you can find [here](https://github.com/pulumi/pulumi-kubernetes-crds/). I still feel like there's additional features to build out in this area though, so I definitely plan to submit a PR here and there even after today.

Reviewer: ```suggestion
Over the course of the summer, I progessively built [`crd2pulumi`](https://github.com/pulumi/crd2pulumi), a CLI tool that generates a typed CustomResource from a CRD. I then leveraged this tool to generate libraries for various Kubernetes Operators, which you can find [here](https://github.com/pulumi/pulumi-kubernetes-crds/). I still feel like there's additional features to build out in this area though, so I definitely plan to submit a PR here and there even after today.
```

In contrast, real software development at Pulumi was far more multi-person, rich, and dynamic. It was annoying at times, but also far more rewarding. In the first couple of weeks, I was seemingly running into issues at all layers—with the `pulumi/pulumi` codegen, with the Go compiler, with the Pulumi CLI, with the Python SDKs, or with my Macbook overheating and crashing whenever I tried opening up the `examples` repository in VSCode. Getting stuff to "just work" often required fixing many nested sub-problems until the root problem could finally be resolved. After a while, I internalized a lot of general debugging heurstics, and developed better skills in squashing bugs and getting code to "just work."

I thought that Pulumi gave me an excellent balance between guidance and autonomy. My co-workers gave me an outline on what my project should acheive, and some pointers on how which tools to use, but other than that it was up to me to implement it.

> One aspect that I liked about Pulumi was that its main audience is of course developers, so it was easier to design `crd2pulumi` since I myself am a developer. Being able to see and directly interact with other programmers that use my tool and offer feedback was immensely satisfying. I loved the energy that came from working with a fast-growing startup that provided real value to real people. It was exciting to come to Monday's team meeting and see Pulumi's user count contiously rise, and hear about the latest deal that the sales team had just closed.

Reviewer: Typo: contiously => continuously

```suggestion
One aspect that I liked about Pulumi was that its main audience is of course developers, so it was easier to design `crd2pulumi` since I myself am a developer. Being able to see and directly interact with other programmers that use my tool and offer feedback was immensely satisfying. I loved the energy that came from working with a fast-growing startup that provided real value to real people. It was exciting to come to Monday's team meeting and see Pulumi's user count continuously rise, and hear about the latest deal that the sales team had just closed.
```

> One aspect that I liked about Pulumi was that its main audience is of course developers, so it was easier to design `crd2pulumi` since I myself am a developer. Being able to see and directly interact with other programmers that use my tool and offer feedback was immensely satisfying. I loved the energy that came from working with a fast-growing startup that provided real value to real people. It was exciting to come to Monday's team meeting and see Pulumi's user count contiously rise, and hear about the latest deal that the sales team had just closed.

Reviewer 2: Add parenthetical commas (`of course` -> `, of course,`)

```suggestion
One aspect that I liked about Pulumi was that its main audience is, of course, developers, so it was easier to design `crd2pulumi` since I myself am a developer. Being able to see and directly interact with other programmers that use my tool and offer feedback was immensely satisfying. I loved the energy that came from working with a fast-growing startup that provided real value to real people. It was exciting to come to Monday's team meeting and see Pulumi's user count contiously rise, and hear about the latest deal that the sales team had just closed.
```

## Pros of the Virtual Life

I hugely appreciated the flexibility of virtual work-from-home. In conjunction with the two summer classes I was taking, this allowed to me to ocassionally optimize my schedule. For example, if I had stayed up late on Tuesday night submitting a problem set and I knew I had no meetings tommorow, I could wake up later on Wednesday morning, but just work a few more hours in the evening to compensate. However I tried my best to avoid this to keep in sync with the majority of Pulumi employees that are on from 9am to 5pm PST.

After a few days, I settled into a routine of:

> - Wake up around 9-10am and drink some water

Reviewer: Blank line before bullets to make the linter happy

```suggestion

- Wake up around 9-10am and drink some water
```

- Fire up my Macbook, check Slack and emails, attend meetings if scheduled, read through the reviews to my PR and write some code

> - Eat brunch around 1-2pm (almost always a double-wrapped Chipotle burrito with brown rice and chicken)

Reviewer 2: Interesting lunch choice! I didn't know "double-wrapped" was a thing at Chipotle. You'll have to explain how that works at the upcoming lunch.

- Write some more code, fix some bugs, push some changes until 5-6pm

In addition, I feel that the quality of the technical practice I received would have been the same, had the internship been in-person. I usually knew which feature or fix to accomplish, and could troubleshoot most technical errors along the way. If I did run into a bug with my code or had questions on how to implement some function, I could always message a co-worker or setup a meeting to resolve it pretty quickly. I also picked up on a lot of knowledge regarding Pulumi, our tech stack, and best practices by just listening in on meetings and browsing the company Slack channel.

> ## Cons of the Vritual Life

Reviewer: ```suggestion
## Cons of the Virtual Life
```

For me, the largest drawback to work-from-home was the diminished social opportunities. Everybody I met at Pulumi was so interesting, talented, and kind, so I'm disappointed that I couldn't meet them in person. There's no online replacement for actually seeing a person for 8 hours a day and 5 days a week. I feel that so much of life is disproportionately formed via one-off, chance encounters with people, so not being able to go down to a city such as Seattle and work in an office limited my [exposure to the envelope of serendipity](https://nevalalee.wordpress.com/2011/10/16/nassim-nicholas-taleb-on-maximizing-serendipity/). Looking back, I should've set up more 1-1 meetings with people at Pulumi, just to chat with them and learn from their experiences over lunch.

> Nevertheless it was still enjoyable, and Pulumi tried its hardest to keep us engaged with remote social activites. For example, [Lee](https://www.pulumi.com/blog/author/lee-zen/) set up a couple of escape room games which were really entertaining. 

Reviewer: Relative link and trailing whitespace (the linter was complaining)

```suggestion
Nevertheless it was still enjoyable, and Pulumi tried its hardest to keep us engaged with remote social activites. For example, [Lee](/blog/author/lee-zen/) set up a couple of virtual escape room games which were really entertaining.
```
